pkg/exchange/bitget/types.go

```diff
@@ -0,0 +1,149 @@
+package bitget
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+type InstType string
+
+const (
+	instSp InstType = "sp"
+)
+
+type ChannelType string
+
+const (
+	// ChannelOrderBook snapshot and update might return less than 200 bids/asks as per symbol's orderbook various from
+	// each other; The number of bids/asks is not a fixed value and may vary in the future
+	ChannelOrderBook ChannelType = "books"
+	// ChannelOrderBook5 top 5 order book of "books" that begins from bid1/ask1
+	ChannelOrderBook5 ChannelType = "books5"
+	// ChannelOrderBook15 top 15 order book of "books" that begins from bid1/ask1
+	ChannelOrderBook15 ChannelType = "books15"
+)
+
+type WebsocketArg struct {
+	InstType InstType    `json:"instType"`
+	Channel  ChannelType `json:"channel"`
+	// InstId Instrument ID. e.q. BTCUSDT, ETHUSDT
+	InstId string `json:"instId"`
+}
+
+type WsEventType string
+
+const (
+	WsEventSubscribe   WsEventType = "subscribe"
+	WsEventUnsubscribe WsEventType = "unsubscribe"
+	WsEventError       WsEventType = "error"
+)
+
+type WsOp struct {
+	Op   WsEventType    `json:"op"`
+	Args []WebsocketArg `json:"args"`
+}
+
+// WsEvent is the lowest level of event type. We use this struct to convert the received data, so that we will know
+// whether the event belongs to WsOpEvent or WsChannelEvent.
+type WsEvent struct {
+	Arg WebsocketArg `json:"arg"`
+	// "op" and "channel" are exclusive.
+	*WsOpEvent
+	*WsChannelEvent
```

Comment:
embed multiple struct like this may be dangerous?

Reply:
I think that is fine because we have a sanity check before calling the pointer struct. 
```
func (w *WsEvent) IsOp() bool {
	return w.WsOpEvent != nil && w.WsChannelEvent == nil
}

func (w *WsEvent) IsChannel() bool {
	return w.WsOpEvent == nil && w.WsChannelEvent != nil
}
```

This event is only used in parseWebSocketEvent. If it does not belong to any type, it will return an error.
```
func (s *Stream) parseWebSocketEvent(in []byte) (interface{}, error) {
	var e WsEvent

	err := json.Unmarshal(in, &e)
	if err != nil {
		return nil, err
	}

	switch {
	case e.IsOp():
		return e.WsOpEvent, nil

	case e.IsChannel():
		switch e.Arg.Channel {
		case ChannelOrderBook, ChannelOrderBook5, ChannelOrderBook15:
			...
			return &books, nil
		}
	}

	return nil, fmt.Errorf("unhandled websocket event: %+v", string(in))
}
```